[FEATURE] Publish stable types for Ember

chriskrycho opened the pull request:
🎉🎉🎉🎉🎉

***This PR does exactly what it says on the tin: publishes stable types from Ember's own TypeScript source code!***

🎉🎉🎉🎉🎉

We started working in earnest on writing types for Ember over 6 years ago, in early 2017; we started down the official "Road to TypeScript" 2 full years ago, in early 2021; and I started work on the mechanics we use to publish these types last year. This is one of the last major blockers for implementation for emberjs/rfcs#800, allowing us to have full TypeScript support and it'll make a nice headliner feature for Ember 5.1!

## SemVer

**Note on SemVer:** assuming these types are included in the Ember 5.1 release, we are not *thereby* committed to treating them as stable. They should be treated as stable once we start opting users into them automatically, by including the appropriate imports in the blueprints. That said, we should *plan* to do exactly that, and follow our normal SemVer commitment, as elaborated in [the spec we wrote and adopted for TS types](https://www.semver-ts.org).

## Implementation details

Mechanically:

- Makes a number of small but important updates to the types publishing script, to enable it to handle more edge cases.
- Fixes a number of type definitions internally, when and where the type tests exposed problems with them.
- Adds a number of new types to Ember's internals (and at least "intimate" APIs) which allow it to properly integrate with Glint, using the same mechanics as we used in the DT and preview types packages.
- Removes the `preview` types package entirely, since it is now defunct.
- Updates the type tests themselves to match the implementation where the preview types (previously tested against) were wrong.

Notes on the implementation/for reviewers:

1. There are many commits, and the number of changes here is not small, but it should be comprehensible by working through individual commits, because I have done my very best to make each commit atomic and meaningful… but there are some caveats to that. Because this was an iterative process, there are a few cases where I made a change and then came back to it later in a significantly different way; the key one of these to note is around the `.create()` and `.extend()` types, where the final output is somewhat different than you might think from the initial commits addressing it.

2. I opted to strike a balance between *internal utility* and *external maintainability* for the types we use for the classic class system. In most cases, this had a minimal impact on our internal usage. Notably, it means we do provide *slightly* more support than we originally promised in the RFC. On balance, I think this is fine; in general, we will still strongly nudge people toward native classes because the support is finicky at best. This was the only way to preserve our reliance on some of those features internally without writing unsafe casts for the external types, so it seemed best.

3. I would have liked to break this up into smaller pieces at a time, and indeed did so with what chunks I could last year. However, after taking four or five different passes at that, I realized that there is far, far too much coupling for that to be tractable as things stand. We made some good progress on the internal coupling last year, and I expect that it will become increasingly tractable over 5.0 and, if we deprecate the Classic mechanics during v5, after their removal at v6.0. For now, however, things are too messy, so the only way to ship *any* of the remaining types was to ship *all* of them.

4. This publishes *all* the types: internal and external alike. It would be nice to have a more robust module structure which does not require us to do that, but that is a significantly larger lift. This at least affords us the ability to publish our types!

## Thanks

- LinkedIn sponsored this work—I've spent about a quarter of my time over the last 12 months preparing for this and then delivering on it.
- @dfreeman helped me repeatedly when stuck on extremely thorny type issues, and he and @jamescdavis have helped drive the TS-in-Ember project forward alongside me end to end.
- @cafreeman and @rwjblue provided great sounding boards, feedback, and input on details along the way.

chriskrycho committed:
Change how loader's `define` becomes visiblte to TS

Previously, this was defined as a global in `ember-template-compiler`.
This was not really correct; it was a workaround for the way that the
`loader` package defines it as a global as a side effect of importing
and executing the module. Here, make it visible specifically *and only*
to the root `ember` package, where it is needed for defining the type
of the `Ember.__loader` object.

chriskrycho committed:
Provide an `owner-ext` for `@ember/service`

`owner-ext` is a pass-through type definition module which adds the
`Registry` from `@ember/service` to the `DIRegistry` in `@ember/owner`
in a way that works for both internal usage within Ember *and* usage by
end users, given the constraints of our types publishing infrastructure
and its need to wrap all proper modules in `declare module { }` blocks.
Since this already *has* a `declare module { }` block, it does not need
to be wrapped; and since it is simply a `.d.ts` file, we can copy it
over rather than trying to emit it with `tsc` during build. (This is,
admittedly, a hack: but a sadly necessary one given our constraints!)

chriskrycho committed:
Import a root without `/index` in `ember-template-compiler`

Instead of working around this in the type publishing infrastructure
(though we may ultimately want to do that as well), just *fix* it by
normalizing the import to the more "normal" form without `/index`.

chriskrycho committed:
The router service presently takes a generic

We should, separately, fix that: it doesn't actually make sense for the
`RouterService` to be generic over a single route or even a set of
routes, so we need to go back and undo that type-level plumbing.

chriskrycho committed:
Update how we compile backburner.js' types

Use the full package name in both the broccoli build and all imports
which reference it, so that the types can resolve to them correctly.

chriskrycho committed:
Publishing script: do not wrap passthrough copies

This makes it so we can write code for things like the `DIRegistry`
which works correctly at both authoring and publishing time: it means
the side-effect-type modules can be in "module mode" rather than purely
ambient mode in both contexts. It also updates the definition of the
`@ember/service/owner-ext.d.ts` to be in module mode.

chriskrycho committed:
Correct the `@ember/application` type test

The previous test was overly generous, allowing users to pass anything
into the `buildInstance()` call and not checking that it worked with
the actual required arguments. This may technically be true, given how
the Ember APIs for `.create()` work under the hood (and that is how
this works, of course), but it's not *useful* to TS users, and it does
not actually represent the expected public API.

chriskrycho committed:
Do not special-case ember-template-compiler/lib/types

These can go through the normal pipeline by being a `.ts` files instead
of a `.d.ts` file in the source.

chriskrycho committed:
Use `keyof this` in `Observable`'s `get`, `set`, etc.

This makes `this.get('foo')` and friends work in roughly the same cases
where `get(this, 'foo')` and friends work.

chriskrycho committed:
Emit named items for built-in components

The `Input` and `Textarea` components were previously being emitted as
the direct result of the `opaquify` item, with no name. This made it
impossible for TS to *automatically* emit them (a) with a useful import
name or (b) with a nameable type. Introducing an intermediate binding
and a type defined as the `typeof` the named constant, the combination
of which becomes the default export, maintains the existing runtime
behavior exactly while making the types name-able.

chriskrycho committed:
Correctly handle router registration into service registry

Use the same pattern for this as for extending the DIRegistry with the
service registry: create a passthrough copy of a module declaration
which is valid in both authoring and publishing contexts.

chriskrycho committed:
Name the return type of (internal) controllerFor

This prevents us from needing to emit (rewritten) relative paths for
the return type.

chriskrycho committed:
Loosen restriction on service registry

Services can be any object, both in practice for today and in a future
where Ember services don't have to have an Ember base class.

chriskrycho committed:
Substantially refactor `publish.mjs`

Introduce the ability to copy over hand-authored `.d.ts` files; Ember
has a number of these for things which exist at runtime but have no
corresponding authoring-time `.js`/`.ts` file, e.g. templates.

Along the way, make a couple significant improvements:

1. Switch to using the async versions of most operations, which mildly
   improves performance.
2. Extract some of the functionality into well-scoped functions to show
   the flow more clearly, and to handle the common process of writing
   an error to the console and bailing.

chriskrycho committed:
Fix types and type tests for `@ember/array`

- `sortBy()` can accept multiple keys to sort on
- Type tests should not name `EmberArray` as `Array` lest they confuse
  the difference with global (real!) `Array`.
- Type tests should use public API imports, not private API imports.

chriskrycho committed:
Publish types from source and remove preview types

Remove *all* preview types, update the publishing script to publish all
packages from source, and update the publishing tsconfig to include all
source packages.

**This *does not work* yet.** However, it is a necessary first step in
the process of switching to publishing stable types: the remaining
types are too deeply entangled with each other to publish them in
chunks as we have to date. (If there are exceptions, I have not found
them, and the overhead of finding them is higher than the cost of just
doing this switch-over.)

chriskrycho committed:
Add ability to manually copy `.d.ts` type modules

Some packages in the project supply `.d.ts` files which must be part of
both the authoring-time types *and* the published types, because they
represent values which become available at runtime by way of codegen.

For example, the `ember/version` module provides a `VERSION` value
derived from the package version. Both Ember's internals and the public
API include this value in various ways, but it does not have a
corresponding ES module on disk, so we provide a type definition
representing where it *notionally* lives, and that type definition must
also appear in the published types.

In these cases, simply copy the file from an input to an output
location, rather than trying to use `tsc` for them (since `tsc` will
actually ignore them).

chriskrycho committed:
Correct `@ember/object` type tests given improved safety

Previously (i.e. on earlier versions of TS) we did not consistently get
the type params preserving the underlying types as robustly, but we can
now preserve this information.

chriskrycho committed:
Update `@ember/application` type tests for DI registry

Now that the service registry provides a constraint of `object` for the
entries, rather than no constraint whatsoever, we can update the tests
to verify this to prevent regression.

chriskrycho committed:
Handle type (un)safety of some runloop `bind`

We can safely type *almost* all variants, but not using Ember's `bind`
with `this` and a string name. Introduce an appropriate fallback for
that case; this means that we cannot catch quite as many errors but in
exchange we get coverage over documented patterns in Ember's code.

(This gives us the same behavior as the preview and DT types always had
as a fallback.)

chriskrycho committed:
Handle `declare namespace` in types publishing script

`namespace` and `module` both introduce TS `ModuleDeclaration` types,
so we distinguish between the two by checking whether it uses a string
literal (rather than an identifier) when excluding `declare module`
from the post-processing.

chriskrycho committed:
Refactor `Ember` namespace to be a TS `namespace`

While the `namespace` construct is very much to be avoided for new
code, it is *exactly* correct for the case of the `Ember` namespace,
which is (precisely!) a namespace. To make this work:

- Combine the existing `const PartialEmber` and `const Ember` object
  bindings into a single `namespace Ember`.

- For the members assigned via `Object.defineProperty`, switch from an
  `interface` to `export declare let` and `export declare const`
  bindings on the `namespace`. Make sure to use `let` or `const` to
  match the readonly or writable semantics of the items defined thus.

Testing this exposed that at some point `Ember.run.later` and similar
nested namespace access stopped working, almost certainly at the point
when we dropped support for the namespace access from the module, but
our type tests still promised it. Noting that no *runtime* tests check
for this, remove the corresponding type tests.

chriskrycho committed:
Add top-level safe key list for `getProperties`

This already works in the variant where the caller passes in an array,
but we did not have the overload for where they pass in multiple args.

chriskrycho committed:
Provide a useful type for `@computed.meta()`

Useful and *hopefully* accurate, but the implementation is a bit murky
as to what the intended semantics are, and the commit history is not
illuminating. This seems to capture the actual tested and documented
behavior accurately.

chriskrycho committed:
Extract owner and location side effect modules for routing/location

As with our service registry, we need these to live in separate files
so they can be copied over without change and so be able to perform
imports while *also* performing a module merge.

chriskrycho committed:
Dramatically simplify routing types, fix its type tests

Since we have no ability to connect or check, within Ember itself or in
the things it provides to callers, the relationship between any given
Route model, Transition, and the overall RouterState, remove those type
parameters entirely in favor of just passing Ember's `Route` as the
type directly to `router.js`'s type parameters. This lets us simplify
the routing layer types considerably.

Also update a number of the type tests for the routing layer to match
Ember's actual behavior: the preview (and DT) types were simply wrong
at various points.

chriskrycho committed:
Make cache-for slightly type-safer

Add an overload which correctly handles keyof lookups.

chriskrycho committed:
Update types and tests for Ember namespace

- Give onerror a useful type (the one it has always had).
- Mark some exported items as private.
- Update the test suite to reflect the types from source: in some cases
  looser, in other cases somewhat tighter!

chriskrycho committed:
Remove extraneous type parameter from ArrayProxy

chriskrycho committed:
Make the `compare` utility more type safe

chriskrycho committed:
Correctly capture the overload semantics of `warn`

chriskrycho committed:
Fix type errors (by using simpler types!) for utils

chriskrycho committed:
Fix types and type tests for `@ember/debug/data-adapter`

- Introduce `Column` type and use it as appropriate.
- Update type tests for the *actual* API: apparently DT had this wrong
  for many years.

chriskrycho committed:
Update type tests to match RouterService implementation

chriskrycho committed:
Fix types and type tests for @ember/debug

- `deprecate` and `warn` can be called without args (this is not the
  intended behavior, but it is currently legal in the internals).
- `assert` needs to be typed correctly in the implementation to be
  `never` when no predicate is passed.

chriskrycho committed:
Improve local types for component-managers

chriskrycho committed:
Fix types for `assert` and `(get|set)DebugFunction`

- Update prod `noop` bindings to cast away safety.
- Update type tests to expect `assert` to `asserts test` or be `never`.

chriskrycho committed:
Explicitly name an indexed type in router.ts

chriskrycho committed:
Set correct types for QUnit test adapter

chriskrycho committed:
Allow `ember-testing`'s `Promise` not to have labels

chriskrycho committed:
Add a required `as const` for Controller QPs type test

TS helpfully tries to infer a mutually-compatible type when passed more
than one object in an array (to allow narrowing etc.), but with QPs on
a Controller, we actually do *not* want that behavior; so when passing
both `type` and a rename in the same QPs config, use `as const`.

chriskrycho committed:
Supply correct internal types for `precompileTemplate`

Although this is a build-time-only API, which is ultimately removed
from the distributed build, it is perfectly legal for Ember users to
*name*, especially if doing their own build tooling work; as such, it
needs to provide the types it has when used in the build.

chriskrycho committed:
Handle edge case in types publishing path mapping

Without this, some TS-generated `import()` types naming a given item
where the runtime code simply relies on inference get emitted without
the top-level `@ember/` namespace. This special case makes sure we do
not drop that: if we ever find it, we know we *need* to emit it.

chriskrycho committed:
Update the Ember NS `debug` tests to match modules

chriskrycho committed:
Update types and type tests for CP macros

- Correctly capture deprecation options.
- Flag CP syntax we do not expect to work.
- Make `@sort` less type-safe, to match `Array.prototype.sort`.

chriskrycho committed:
Simplify types for `(get|set)Properties`

This fixes many of the incompatibilities from our type tests, and makes
the resulting code *much* clearer at the same time. The key here is to
move away from trying to use tuple spread types to just constraining
the actual args and mapping their types into a `Pick` or `Record`.

chriskrycho committed:
Account for optionality of `assertDestroyablesDestroyed`

chriskrycho committed:
Eliminate cycle in definition of (Ember) `Component`

`Component` uses `ViewSupport`, but `ViewSupport` defined its `this` as
`Component`. This was a classic form of cycle which blocked supporting
the `Mixin`-based programming model effectively with TS in the first
place and was a major motivator for investing in native classes back in
the early Ember 3.x days! here, the fix is to use `this: CoreView`
instead so that `ViewSupport` does not need to name `Component`, only
the chunk of `Component`'s base types it actually uses.

As a bonus, remove a needless `extends FrameworkObject` from the merge
on the interface of the `CoreView` class, since normal inheritance does
correctly preserve that.

chriskrycho committed:
Fix/simplify types of `create`, `reopen`, and `reopenClass`

As described in the RFC, we do not want to attempt to provide the type-
modifying capabilities of these functions. The result is type safe but
not especially useful, but that is as we want: using `create` to
provide values not on the base type *or* to change its values is quite
undesirable; and using `reopen` and `reopenClass` to dynamically update
instance and static members of a class has serious (and not obvious!)
performance downsides and discoverability problems for maintainers, and
will ultimately be deprecated entirely as part of removing Classic.

A key result is that `.create()` now allows *any* props to be set, but
does not update the resulting type. This is an unfortunate necessity
given how Ember *itself* uses `.create()`.

chriskrycho committed:
Update engine type tests now that types catch more errors

chriskrycho committed:
Remove now-defunct types for `(add|remove)Listener`

Ember's own types are now less generous about `evented`'s `addListener`
and `removeListener` functions than the preview/DT types were, because
Ember's implementation has been simplified and the types had not been
updated to match.

chriskrycho committed:
Accept re-mappings in Controller QP type definitions

chriskrycho committed:
Remove extranes `extends` from `EmberObject` interface

Again, this already correctly `extends CoreObject` because the class
itself does so; we only need to merge in declarations with mixins.

chriskrycho committed:
The type of `objectsAt` is `Array`, not `Ember.Array`

chriskrycho committed:
Handle optionality of `Ember.Test` in type tests

chriskrycho committed:
Provide type safe get and set in ObjectProxy

While people should generally move away from this and to native `Proxy`
this makes it possible for existing code to be well-typed.

chriskrycho committed:
Use `Owner` correctly in `@ember/object` type tests

chriskrycho committed:
Fix `@ember/object` tests for proxy and observable

Account for both the improved `Proxy` type safety and the removed
`CoreObject.create` type safety in the type tests.

chriskrycho committed:
Fix bad import and reference in ember NS type tests

These referred respectively to imports which existed only in DT/preview
types and a removed API!

chriskrycho committed:
Partially reimplement type-safe `.create()`

This gets us a *small* amount of type safety, rather than absolutely
none: it does not check the types passed into `.create()` match, but
since there is little reason to directly call this for end user code
(vs. for internal code, which actually uses this dynamic capability!),
this gives us the actual runtime behavior: if you smash a property with
a new type, you... do.

chriskrycho committed:
Correctly export both sides of the opaque LinkTo

Even though we have an "opaque" class here, we still need to make sure
both the static type of the class and its runtime type are available
to callers to name.

chriskrycho committed:
Update a `set()` type test: it allows dynamic setting

This matches the runtime behavior. We do not catch setting things this
way, as a result, but in a world where native property access works
everywhere, this is actually somewhat useful: it means you can set
anything on a target object, and although the type will not be updated
with that change, it will at least type check in a way that matches the
dynamic runtime behavior.

chriskrycho committed:
Remove a type test which only covered internals

This was useful to prevent us from regression in the DT/preview types
era, but these are covered by Ember's internals being required to type
check as part of CI now, and we have also simplified or removed most of
this machinery.

chriskrycho committed:
Update type tests for improved `set`/`get` inference

The resulting types are just records, where `Pick` preserves some extra
info because of the dynamic class behavior.

chriskrycho committed:
Drop `Ember.get` from Ember Router tests

The fact that `Ember.get` gets very confused on deeply mapped types
combined with `this` types is not surprising, but is also not relevant
to a test where no one *should* use `get()` instead of normal property
access.

chriskrycho committed:
Further remove generic Route param from router

As with the earlier commit removing some of this, we have no ability to
check these types, so they are just extraneous overhead.

chriskrycho committed:
Fix types for internal-test-helpers

Now that `create` can catch this, it lets us know about our own lies.

chriskrycho committed:
Finish updating internals for safe-r create

chriskrycho committed:
Improve correctness and usability of `@ember/modifier` types

- Re-export the relevant types from `@glimmer/interfaces` as well as
  wiring them up to the appropriate points in the exported functions.
- Add the hooks necessary to make the exported `on` modifier visible to
  Glint, including adding an `Opaque` internal type.

chriskrycho committed:
Improve types and type tests for `@ember/test`

- Provide the correct type for the first param for the callback passed
  to `registerHelper()`.
- Add a cast and an explanation for type test on `TestAdapter`

chriskrycho committed:
Fix some missed type tests for `.extend`, `.create`, and `.reopen`

chriskrycho committed:
Correctly handle opaque LinkTo type in type tests

chriskrycho committed:
Fix type test for Ember.Object: correct `Owner`

chriskrycho committed:
Correct type tests for `typeOf` (just `string`)

chriskrycho committed:
Add `static positionalParams` type declaration to EmberComponent

Without this, neither `.extend()`ing `Component` nor the documented
`reopenClass` pattern works with `positionalParams`.

chriskrycho committed:
Fix types of ArrayProxy.content

chriskrycho committed:
Fix types around component templates and TOCs

- Make `templateOnlyComponent` and associated types usable with Glint
  the same way the preview types were.
- Fix the way we name the types of `setComponentTemplate` in tests. In
  the real world, the way to get the type is via calling another public
  API; here we just want to make sure the function *requires* the
  correct thing.

chriskrycho committed:
Fix types and type tests for `@ember/test`

- Do not emit `| () => never` for functions in `@ember/test`: while it
  is true that those *will* error if `ember-testing` is not included,
  the underlying function type is still correct even there, since the
  immediately-throwing handler is a legitimate function subtype of the
  expected behavior when the library *is* included. Additionally, that
  inclusion is normally a build-time concern.

- Apply some formatting tweaks to the definition of `registerWaiter`'s
  implementation definition and `prettier-ignore` those lines to make
  it more legible.

- Update a `register` call in a type test to use a proper 'full:name'.

chriskrycho committed:
Drop an unused import from Ember type tests

chriskrycho committed:
Make Ember's helper types Glint-compatible

Provide the same public API as the preview types. This requires lying a
bit about the return type of `helper()` in particular, but no part of
Ember's own internals cares about the details: they are relevant *only*
as the point where they are actually created. The provided type avoids
a number of possible kinds of errors by providing an opaque item which
cannot be constructed but which *does* preserve generics. Moreover, the
actual details of type returned by `helper()` are *not* public API, so
this correctly avoids exposing those details.

As a bonus, remove a number of extra internal type aliases on both the
source and that were used in the preview types.

chriskrycho committed:
Fix optionality of `owner` arg for `Route` class

chriskrycho committed:
Fix types for template-only components

- Do not expose Glimmer's private/internal notion of a template-only
  component, even for type tests.
- Do provide the full *call* API for Glimmer's `templateOnly()`.

chriskrycho committed:
Fix re-export of `Ember.RSVP` as a `namespace`

This syntax is namespace-specific: `import` in a namespace is aliasing
one namespace to another, while `export` marks the item public on the
namespace (as with the rest of the exported items).

chriskrycho committed:
Drop now-unnecessary `types/preview` from type-tests

chriskrycho committed:
Fix TS lint errors

chriskrycho committed:
Remove assertion from build-owner internal test helper

At the moment, some *tests* pass no resolver, and expect this to work
anyway. The fix is to have those tests pass a test-friendly resolver
and then this will work as expected.

chriskrycho committed:
Fix a missed rename in `ember/index.ts`

Supporting the TS `namespace` export for the `Ember` namespace object
(in 2fdb3f57) required aliasing many of the imports; one of these was
hidden by having a `@ts-expect-error` on that statement for unrelated
reasons.

chriskrycho committed:
Update Ember NS reexport test for backburner.js

The existence of the `_Backburner` re-export is of a private item, and
this test simply checks that the re-export works as expected; sinc we
updated the build pipeline not to do the weird renaming shenanigans it
did previously to make `'backburner'` the import downstream, we can
update this test to match the actual Node-resolveable location as well
(see 80b738d2 for background).

chriskrycho committed:
Add private runloop backburner re-exports

`@ember/test-helpers` uses these as they exist in the DT and preview
types, so to avoid breaking that, re-supply them here. They should not
be relied upon long-term, but this allows us to migrate the library's
dependency chain distinctly from the release of these types.

chriskrycho committed:
Remove an unnecessary type cast from Router

dfreeman commented on packages/@ember/-internals/utility-types/index.ts:
```diff
@@ -19,3 +19,12 @@ export type MethodParams<T, M extends MethodNamesOf<T>> = Parameters<MethodsOf<T
 export type OmitFirst<F> = F extends [any, ...infer R] ? R : [];
+
+// A way of representing non-user-constructible types. You can conveniently use
+// this by doing `interface Type extends Data<'some-type-name'> { ... }` for
```
```suggestion
// this by doing `interface Type extends Opaque<'some-type-name'> { ... }` for
```

dfreeman commented on type-tests/@ember/component-test/helper.ts:
```diff
@@ -153,17 +133,22 @@ const badReturnSig = helper<DemoSig>(([i18nizer], { name, age }) =>
+      Named: object;
+    };
+    Return: string;
+  }>
+>();
 
 // @ts-expect-error
 new greet();
 
 // @ts-expect-error
 class Subgreet extends greet {}
 
+// Check that generics are accepted. (We cannot meaningfully test for how they
+// are *handled* because it uses items we do not want to expose in public API.)
 const pair = helper(<T>([item]: [T]): [T, T] => [item, item]);
-expectTypeOf(pair).toEqualTypeOf<
-  abstract new <T>() => FunctionBasedHelperInstance<{
-    Args: { Positional: [T]; Named: EmptyObject };
-    Return: [T, T];
-  }>
->();
```
What will TS produce when emitting declarations for a helper like this if `FunctionBasedHelperInstance` isn't exported?

chriskrycho replied:
Did some poking and confirmed that we do need this type exported. @dfreeman and I spent a bunch of time discussing alternatives, but ultimately landed on "just export this thing and look forward to when the need for it goes away by dropping `helper()` in favor of *just* using plain functions instead"; done in c725cdce7.

dfreeman commented:
I left a few comments/questions, and am hoping to get the Glint monorepo typechecking against a build from this branch tomorrow to make sure everything works cleanly there, but overall this looks fantastic! 🎉

chriskrycho commented on packages/@ember/-internals/glimmer/lib/components/input.ts:
```diff
@@ -270,4 +270,6 @@ class Input extends AbstractInput {
 
-export default opaquify(Input, InputTemplate);
+const Input = opaquify(_Input, InputTemplate);
+type Input = typeof Input;
```
Ah, yes:

https://github.com/emberjs/ember.js/blob/b1ebc7d4abf99b04168e824137d8e2ecd3a329a6/types/preview/@ember/component/index.d.ts#L145-L146

I'll update all of these to match that. Thanks!

chriskrycho commented on types/publish.mjs:
```diff
@@ -97,94 +96,136 @@ ${MODULES_PLACEHOLDER}
+    {
+      input: 'packages/@ember/routing/service-ext.d.ts',
+      output: '@ember/routing/service-ext.d.ts',
+    },
```
I don't think you're missing anything; I think this is what happens when the authoring of a given piece of infrastructure gets chopped up over the course of many months by massive conflagrations in unrelated areas. 😂 I'll see if I can combine them!

chriskrycho replied:
Fixed via 30a80963e! Separated and improved them.

chriskrycho committed:
Fix typo in comment in broccoli/packages.js

Co-authored-by: Dan Freeman <dfreeman@salsify.com>

chriskrycho committed:
Do not duplicate work on hand-written `.d.ts` files

There are three categories of `.d.ts` files we manually copy over as
part of our publishing process:

- `*-ext.d.ts` modules, which supply module merge declarations manually
  to allow the regular module to go through the normal postprocessing
  chain while still allowing the imports to work correctly for the
  declaration merge. These must be copied over and left exactly as they
  are, no post-processing.

- Modules which correctly represent a runtime-only location of an API,
  which appear at one point in the source graph, but must be emitted at
  a completely different location for runtime/import. These must be
  copied over and left as they are, no post-processing, but targeting a
  different location to emit them.

- Modules which must be copied over and then post-processed, because
  they represent runtime-only code which must be importable as a normal
  module via a relative path *and* must be wrapped in `declare module`
  to be usable in the emitted types.

Separate those pieces of functionality, extract a module-scope list for
the latter two, and rename the local functions accordingly.

chriskrycho commented on packages/ember/index.ts:
```diff
@@ -124,264 +157,281 @@ import type * as EmberTemplateCompiler from 'ember-template-compiler';
-  Component,
+  export const _Input = EmberInput;
+  export const Component = EmberComponent;
+  export type Component = EmberComponent;
```
Ah, it definitely should!

chriskrycho replied:
4bc5d73f

chriskrycho committed:
Include signatures in Ember NS re-exports

chriskrycho committed:
Export built-in components as interfaces for Glint

Instead of `typeof`, export `interface` types which can be merged with
by Glint, and which are kept distinct from each other with the `Opaque`
utility type, while preserving `OpaqueInternalComponentConstructor`.

chriskrycho commented on types/publish.mjs:
```diff
@@ -97,94 +96,136 @@ ${MODULES_PLACEHOLDER}
+
+  - modules which need to live in a different location in the output than in the
+    input tree, e.g. for the loader, which creates runtime modules at a
+    different location than
```
It was! Fixed when I reworked all of this in 22877513.

chriskrycho commented on packages/@ember/routing/route.ts:
```diff
@@ -254,14 +248,11 @@ interface Route<T = unknown> extends IRoute<T> {
 
-  context = {} as T;
-  declare currentModel: T;
+  context = {} as Model;
```
It is not. Unfortunately, it is (apparently?) part of the public API of `router.js`'s `Route` type, so we cannot make it private *here* without also making it private *there*, since this is typed as `implements IRoute` (which is the alias for the one for `router.js`). 😩

chriskrycho replied:
Marked with `@internal` as better-than-nothing signal. b7a7a57c

chriskrycho committed:
Mark internal `Route` details with `@internal`

This does not affect type checking, but since theses items will now be
exposed to consumers, at least explicitly signal via their docs (seen
on hover etc.) that they should not be used.

chriskrycho committed:
Fix typo in description of `Opaque`

Co-authored-by: Dan Freeman <dfreeman@salsify.com>

chriskrycho committed:
Fix typo in comment explaining abstract class for helper

Co-authored-by: Dan Freeman <dfreeman@salsify.com>

chriskrycho committed:
Provide default type for Signatures etc. in Ember NS

Without these, anyone using them directly as a type will get a type
error reporting that they are missing.

chriskrycho committed:
Explicitly type test fallback path for runloop `bind`

This type test will let us know if we change that behavior, rather than
indicating that it is the *desired* behavior.

chriskrycho commented on packages/@ember/runloop/type-tests/bind.test.ts:
```diff
@@ -135,4 +135,6 @@ expectTypeOf(bind(foo, 'test', 1, true, 'baz')).toEqualTypeOf<() => number | voi
 // @ts-expect-error Invalid args
+bind(foo, foo.test, 'string');
+// We would like to catch this, but it ends up in the fallback path.
 bind(foo, 'test', 'string');
```
Good call. Done in 19b31a35e.

chriskrycho committed:
Use `import type` in internal components

chriskrycho committed:
Publicly export `FunctionBasedHelperInstance`

This abstract class -- specifically, its `protected abstract
__concrete__` member -- prevents subclasses from doing `class X extends
helper(..)`, since that is an error at runtime. While it is rare that
people would type that, it is not impossible and we use this to give
them early signal via the types for a behavior which will break (and in
a somewhat inscrutable way!) at runtime.

This is needful because we lie about what this actually is for Glint's
sake: a function-based helper returns a `Factory<SimpleHelper>`, which
is designed to be "opaque" from a consumer's POV, i.e. not
user-callable or constructible but only useable in a template (or via
`invokeHelper()` which also treats it as a fully opaque `object` from a
type POV). But Glint needs a `Helper<S>` to make it work the same way
as class-based helpers. (Note that this does not hold for plain
functions as helpers, which it can handle distinctly.) This signature
thus makes it so that the item is usable *as* a `Helper` in Glint, but
without letting end users treat it as a helper class instance.

Long-term, this entire construct can go away in favor of deprecating
the `helper()` invocation in favor of using plain functions.

chriskrycho commented:
**Status note:** We have hit a bit of a snag with this (though not an intractable one)!

All of the stuff I built here works—and works correctly—but you can’t actually use it in a consuming app or addon yet, because `ember-source` currently puts `@glimmer/*` as `devDependencies` and pre-builds all of that as part of its build—but that means that those types are unavailable, and they’re inherently part of the types we publish.

(Where possible I have decreased that exposure in public APIs, but it is and will for the foreseeable future be impossible for that exposure to be *zero*. This problem is not going away.)

Net, when you try to use the types published from Ember's source, you end up with an awful lot of these (45 across 28 different files in a fresh addon):

```
node_modules/ember-source/types/stable/ember/index.d.ts:110:55 - error TS2307: Cannot find module '@glimmer/manager' or its corresponding type declarations.

110     const _modifierManagerCapabilities: typeof import('@glimmer/manager').modifierCapabilities;
                                                          ~~~~~~~~~~~~~~~~~~
```

One option here is to update `ember-source` to depend on those as `dependencies` instead of as `devDependencies`, just the way we would for a "normal" addon. This will be a mild bit of “incoherence” in that those dependencies *will* be the ones Ember uses… but the copy in `node_modules` will *not* be the ones you actually get, courtesy of the pre-build. (However, I suspect we may want to do that *anyway* as part of moving toward Ember’s packages being normal packages?)

We don’t yet have the ability to use something like `rollup-plugin-ts` which might pull these dependency in for us "automatically" by way of the types rollup, because we still have too many cycles in our packages (hard blocker) *and* a small number of straight-up lies in the form of code we generate and which has no on-disk representation (more solvable but not trivial).

I am discussing this with other folks on Framework Core and Typed Ember Core, and will update once we agree on a path forward.

chriskrycho committed:
Move `@glimmer/*` packages to `dependencies`

This is one step along the path toward making Ember's `@glimmer/*`
dependencies work as normal dependencies without pre-building them into
an Ember distribution, but it is introduced *here* because consumers
cannot use Ember's types as built from source without types from the
many `@glimmer` packages Ember uses as part of both its internals and
its public APIs.

Moving these to `dependencies` will *not* change the build behavior: we
will continue to pre-build Ember from these dependencies for now, so
the actual output consumers receive should not be altered by this. The
only change from a consumer's POV will be that they *do* get all the
`@glimmer/*` packages as part of their installation of `ember-source`
and therefore will be able to use the types. Later work can continue to
tackle the broader build questions.

chriskrycho committed:
Move backburner.js and router_js to `dependencies`

As with the `@glimmer/*` packages, these must be present for TS users
to be able to use Ember's types as published from source, since those
types explicitly depend on and in some cases directly re-export these.

chriskrycho committed:
Move `route-recognizer` to `dependencies`

As with the other packages moved this way (see immediately preceding
commits), this must be in `dependencies` so that it is present for
consumers to use its types.

chriskrycho committed:
Merge remote-tracking branch 'origin/main' into types-from-source

chriskrycho commented:
After discussing with folks on Friday, we're going to go ahead and merge this. It *should* be free of issues, but we will keep a close eye out as we move forward. This should be present in Ember 5.1! 🎉
